src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -178,20 +301,515 @@ func (s *commitLogSource) ReadData(
+				// In order for a minimum most recent snapshot to be valid, it needs to be for a block that
+				// we actually need to bootstrap for that shard. This check ensures that our algorithm doesn't
+				// do any extra work, even if we're bootstrapping different blocks for various shards.
+				continue
+			}
+
+			if mostRecentSnapshotForShard.CachedSnapshotTime.Before(minMostRecentSnapshot) || minMostRecentSnapshot.IsZero() {
+				minMostRecentSnapshot = mostRecentSnapshotForShard.CachedSnapshotTime
+			}
+		}
+
+		if minMostRecentSnapshot.IsZero() {
+			// If we didn't find a minimum most recent snapshot time for this blockStart, just use the
+			// blockStart as the minimum since we'll need to read the entire commit log in this case.
+			minMostRecentSnapshot = blockStart.ToTime()
+		}
+
+		minimumMostRecentSnapshotTimeByBlock[blockStart] = minMostRecentSnapshot
+	}
+
+	return minimumMostRecentSnapshotTimeByBlock
+}
+
+func (s *commitLogSource) bootstrapShardSnapshots(
+	nsID ident.ID,
+	shard uint32,
+	metadataOnly bool,
+	shardTimeRanges xtime.Ranges,
+	blockSize time.Duration,
+	snapshotFiles fs.FileSetFilesSlice,
+	mostRecentCompleteSnapshotByBlockShard map[xtime.UnixNano]map[uint32]fs.FileSetFile,
+) (result.ShardResult, error) {
+	var (
+		shardResult    result.ShardResult
+		allSeriesSoFar *result.Map
+		rangeIter      = shardTimeRanges.Iter()
+		err            error
+	)
+
+	for rangeIter.Next() {
+
+		var (
+			currRange             = rangeIter.Value()
+			currRangeDuration     = currRange.End.Sub(currRange.Start)
+			isMultipleOfBlockSize = currRangeDuration%blockSize == 0
+		)
+
+		if !isMultipleOfBlockSize {
+			return nil, fmt.Errorf(
+				"received bootstrap range that is not multiple of blockSize, blockSize: %d, start: %d, end: %d",
+				blockSize, currRange.End.Unix(), currRange.Start.Unix(),
+			)
+		}
+
+		// Reset this after we bootstrap each block to make sure its up to date.
+		if shardResult != nil {
+			allSeriesSoFar = shardResult.AllSeries()
+		}
+
+		for blockStart := currRange.Start.Truncate(blockSize); blockStart.Before(currRange.End); blockStart = blockStart.Add(blockSize) {
+			snapshotsForBlock := mostRecentCompleteSnapshotByBlockShard[xtime.ToUnixNano(blockStart)]
+			mostRecentCompleteSnapshotForShardBlock := snapshotsForBlock[shard]
+
+			if mostRecentCompleteSnapshotForShardBlock.CachedSnapshotTime.Equal(blockStart) ||
+				// Should never happen
+				mostRecentCompleteSnapshotForShardBlock.IsZero() {
+				// There is no snapshot file for this time, and even if there was, there would
+				// be no point in reading it. In this specific case its not an error scenario
+				// because the fact that snapshotTime == blockStart means we already accounted
+				// for the fact that this snapshot did not exist when we were deciding which
+				// commit logs to read.
+				s.log.Infof(
+					"no snapshots for shard: %d and blockStart: %d",
+					shard, blockStart.Unix())
```

Comment:
Perhaps best to log this as string?  `blockStart.String()`

Reply:
done